Allow to change RHS table in each join

ranquild opened the pull request:
Epic https://github.com/metabase/metabase/issues/40890
Demo https://www.loom.com/share/42ec7fdd09b34d09a75a3254a04c759f

Now we allow to change the rhs table for a join. If there are suggested join conditions, the join can be updated normally. If not, we store the rhs table temporary and enter our draft mode which was previously used for new join conditions only. When the join becomes complete, we replace the old join with the new one.

The only side effect / downside is that the old join remains in the query until there is at least one new join condition. So if you select another table and don't select the columns, then save or visualize - the query won't be changed.

ranquild committed:
Allow to change the rhs table in a complete join condition

ranquild committed:
Join draft

ranquild committed:
Allow each join to go to draft state and back again

ranquild commented on frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinComplete/JoinComplete.tsx:
```diff
@@ -0,0 +1,183 @@
+  onDraftRhsTableChange: (newTable: Lib.Joinable) => void;
+}
+
+export function JoinComplete({
```
Moved from `Join`

ranquild commented on frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinComplete/JoinComplete.tsx:
```diff
@@ -0,0 +1,183 @@
+    onJoinChange(newJoin);
+  };
+
+  const handleTableChange = (newTable: Lib.Joinable) => {
```
Everything else is unchanged

ranquild committed:
Remove the old test

ranquild committed:
Remove the old test

kamilmielnik commented on frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinComplete/JoinComplete.tsx:
```diff
@@ -0,0 +1,186 @@
+import { useMemo, useState } from "react";
```
User needs to click the text (name of table) to edit RHS table despite `cursor: pointer` shown for the entire button:

https://github.com/metabase/metabase/assets/6830683/b6f695a9-234d-4a6a-ad36-ec54c219098a

ranquild replied:
Good catch, I'll fix that

kamilmielnik replied:
It's still there, I get it very often.

ranquild replied:
@kamilmielnik fixed. I did it in the same way as it is in DataStep

kamilmielnik reviewed (commented):
Looks good, left comments inline.

I also found #40940 while testing this.

ranquild committed:
Pass join

bshepherdson committed:
[MBQL lib] `suggestedJoinConditions` accepts a join position for edits

This optional fourth argument will disregard all joins at `position` and
later when suggesting conditions. This is useful when editing a join's
RHS, which wipes out the condition. We don't want to generate a
suggested condition that's based on keys from the RHS of the join we
just edited, or from later joins.

Fixes #40916. Part of #40890.

ranquild committed:
Merge remote-tracking branch 'origin/mblib-suggested-join-conditions-for-edits' into join-changing-rhs-table

ranquild committed:
FE fixes

ranquild committed:
Fix table picker button

ranquild commented on frontend/src/metabase/query_builder/components/notebook/steps/JoinStep/JoinComplete/JoinComplete.tsx:
```diff
@@ -0,0 +1,186 @@
+      const newJoin = Lib.joinClause(newTable, newConditions);
+      onJoinChange(newJoin);
+    } else {
+      onDraftRhsTableChange(newTable);
```
Should be fixed now

kamilmielnik reviewed (approved):
:+1:

ranquild committed:
Merge remote-tracking branch 'origin/master' into join-changing-rhs-table
